File: python/src/b3_composite.py
"""
The intent is to compose objects into tree structures to represent part-whole hierarchies.
Composite lets clients treat individual objects and compositions of objects uniformly.

Key participants:
1. Component - declares the interface for objects in the composition
2. Leaf - represents individual objects with no children
3. Composite - represents groups of objects that can have children
4. Client - manipulates objects through the Component interface

Key benefits:
- Treats individual objects and compositions uniformly
- Simplifies client code - no need to distinguish between leaf and composite
- Easy to add new kinds of components

Real-world example: A file system where files and directories should be treated uniformly.
Both can be renamed, moved, or have their size calculated, but directories can contain other items.
"""

from abc import ABC, abstractmethod
from typing import Iterator


# ----- Component interface -----
# Declares the interface for objects in the composition
class FileSystemNode(ABC):
    @abstractmethod
    def print(self, indent: str = "") -> None:
        pass

    @abstractmethod
    def size(self) -> int:
        pass

    @abstractmethod
    def get_name(self) -> str:
        pass

    # Composite-specific operations (default implementation for leaves)
    def add(self, node: "FileSystemNode") -> None:
        print("Cannot add to a file")

    def remove(self, name: str) -> None:
        print("Cannot remove from a file")

    def get_children(self) -> Iterator["FileSystemNode"]:
        return iter([])


# ----- Leaf -----
# Represents individual objects (files) with no children
class File(FileSystemNode):
    def __init__(self, name: str, file_size: int) -> None:
        self._name = name
        self._size = file_size

    def print(self, indent: str = "") -> None:
        print(f"{indent}File: {self._name} ({self._size} bytes)")

    def size(self) -> int:
        return self._size

    def get_name(self) -> str:
        return self._name


# ----- Composite -----
# Represents groups of objects (directories) that can have children
class Directory(FileSystemNode):
    def __init__(self, name: str) -> None:
        self._name = name
        self._children: list[FileSystemNode] = []

    def add(self, node: FileSystemNode) -> None:
        self._children.append(node)

    def remove(self, name: str) -> None:
        self._children = [c for c in self._children if c.get_name() != name]

    def print(self, indent: str = "") -> None:
        print(f"{indent}Directory: {self._name}/ ({self.size()} bytes total)")
        for child in self._children:
            child.print(indent + "  ")

    def size(self) -> int:
        return sum(child.size() for child in self._children)

    def get_name(self) -> str:
        return self._name

    def get_children(self) -> Iterator[FileSystemNode]:
        return iter(self._children)

    def child_count(self) -> int:
        """Return the number of children in this directory."""
        return len(self._children)


# ----- Example usage -----
def composite_example() -> None:
    """
    Demonstrates the Composite pattern with a file system structure.
    """
    print("\n--- Composite Pattern Example ---")
    print("Building a file system structure:\n")

    # Create individual files
    file1 = File("readme.txt", 1024)
    file2 = File("main.cpp", 4096)
    file3 = File("utils.cpp", 2048)
    file4 = File("config.json", 512)

    # Create directories and add files
    src_dir = Directory("src")
    src_dir.add(file2)
    src_dir.add(file3)

    config_dir = Directory("config")
    config_dir.add(file4)

    root_dir = Directory("project")
    root_dir.add(file1)
    root_dir.add(src_dir)
    root_dir.add(config_dir)

    # Print the entire structure (treats all nodes uniformly)
    root_dir.print()

    # Calculate total size (works for both files and directories)
    print(f"\nTotal project size: {root_dir.size()} bytes")


# Test helpers
def create_file_system() -> Directory:
    """Helper for tests: Create a sample file system structure."""
    file1 = File("readme.txt", 1024)
    file2 = File("main.cpp", 4096)
    file3 = File("utils.cpp", 2048)

    src_dir = Directory("src")
    src_dir.add(file2)
    src_dir.add(file3)

    root_dir = Directory("project")
    root_dir.add(file1)
    root_dir.add(src_dir)

    return root_dir


def create_nested_directories() -> Directory:
    """Helper for tests: Create deeply nested directory structure."""
    file1 = File("file1.txt", 100)
    file2 = File("file2.txt", 200)

    inner_dir = Directory("inner")
    inner_dir.add(file2)

    outer_dir = Directory("outer")
    outer_dir.add(file1)
    outer_dir.add(inner_dir)

    return outer_dir


if __name__ == "__main__":
    composite_example()

File: rust/src/b3_composite.rs
//! # Composite Pattern
//!
//! The Composite pattern composes objects into tree structures to represent
//! part-whole hierarchies. It lets clients treat individual objects and
//! compositions uniformly.
//!
//! ## Key Participants
//! - **Component**: Declares interface for objects in composition
//! - **Leaf**: Represents leaf objects in composition
//! - **Composite**: Stores child components and implements child operations
//!
//! ## Benefits
//! - Defines class hierarchies of primitive and composite objects
//! - Simplifies client code - treats primitives and composites uniformly
//! - Makes adding new component types easy

// Component trait
pub trait FileSystemItem {
    fn get_name(&self) -> &str;
    fn size(&self) -> u64;
    fn display(&self, indent: usize);
}

// Leaf
pub struct File {
    name: String,
    size: u64,
}

impl File {
    pub fn new(name: &str, size: u64) -> Self {
        Self {
            name: name.to_string(),
            size,
        }
    }
}

impl FileSystemItem for File {
    fn get_name(&self) -> &str {
        &self.name
    }

    fn size(&self) -> u64 {
        self.size
    }

    fn display(&self, indent: usize) {
        let padding = " ".repeat(indent);
        println!("{}File: {} ({} bytes)", padding, self.name, self.size);
    }
}

// Composite
pub struct Directory {
    name: String,
    children: Vec<Box<dyn FileSystemItem>>,
}

impl Directory {
    pub fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            children: Vec::new(),
        }
    }

    pub fn add(&mut self, item: Box<dyn FileSystemItem>) {
        self.children.push(item);
    }

    pub fn remove(&mut self, index: usize) {
        if index < self.children.len() {
            self.children.remove(index);
        }
    }

    pub fn child_count(&self) -> usize {
        self.children.len()
    }
}

impl FileSystemItem for Directory {
    fn get_name(&self) -> &str {
        &self.name
    }

    fn size(&self) -> u64 {
        self.children.iter().map(|child| child.size()).sum()
    }

    fn display(&self, indent: usize) {
        let padding = " ".repeat(indent);
        println!("{}Directory: {}/", padding, self.name);
        for child in &self.children {
            child.display(indent + 2);
        }
    }
}

/// Helper function to create a sample file system
pub fn create_file_system() -> Directory {
    // Create files
    let readme = Box::new(File::new("readme.txt", 1024));
    let main_cpp = Box::new(File::new("main.cpp", 4096));
    let utils_cpp = Box::new(File::new("utils.cpp", 2048));

    // Create src directory
    let mut src_dir = Directory::new("src");
    src_dir.add(main_cpp);
    src_dir.add(utils_cpp);

    // Create root directory
    let mut root_dir = Directory::new("project");
    root_dir.add(readme);
    root_dir.add(Box::new(src_dir));

    root_dir
}

/// Example demonstrating Composite pattern
pub fn example() {
    println!("\n--- Composite Pattern Example ---\n");

    let mut fs = create_file_system();
    fs.display(0);

    println!("\nTotal size: {} bytes", fs.size());
    println!("Child count: {}", fs.child_count());
    println!("Root name: {}", fs.get_name());

    // Demonstrate remove functionality
    println!("\nRemoving first item...");
    fs.remove(0);
    println!("New child count: {}", fs.child_count());
    println!("New total size: {} bytes", fs.size());
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_file_creation() {
        let file = File::new("test.txt", 1024);
        assert_eq!(file.get_name(), "test.txt");
        assert_eq!(file.size(), 1024);
    }

    #[test]
    fn test_directory_can_contain_files() {
        let mut dir = Directory::new("test_dir");
        dir.add(Box::new(File::new("file1.txt", 100)));
        dir.add(Box::new(File::new("file2.txt", 200)));
        assert_eq!(dir.child_count(), 2);
    }

    #[test]
    fn test_nested_directory_structure() {
        let mut outer = Directory::new("outer");
        let mut inner = Directory::new("inner");
        inner.add(Box::new(File::new("file1.txt", 100)));
        outer.add(Box::new(inner));
        outer.add(Box::new(File::new("file2.txt", 200)));

        assert_eq!(outer.child_count(), 2);
        assert_eq!(outer.size(), 300);
    }

    #[test]
    fn test_calculates_total_size() {
        let fs = create_file_system();
        // readme.txt (1024) + main.cpp (4096) + utils.cpp (2048)
        assert_eq!(fs.size(), 7168);
    }

    #[test]
    fn test_helper_function() {
        let fs = create_file_system();
        assert!(fs.child_count() > 0);
        assert!(fs.size() > 0);
    }
}
